### ISSUE TITLE
Duplicate exports via export * which map to the same externalized implementation disappear from output

### ISSUE THREAD

**Issue by dzearing** [evanw/esbuild#2658]
Repro (not this is not specific to a library, just using my case as an example):

`index.js`: `export * from './path1.js'; export * from './path2.js';`
`path1.js`: `export { DateTime } from '@fluentui/date-time-utilities';`
`path2.js`: `export { DateTime } from '@fluentui/date-time-utilities';`

In this setup, the single `index` barrel file exports both `path1` and `path2`. Each of them export a single named export from a library. This does mean that `DateTime` is exported twice, but resolves to the same symbolic representation.

This works fine in a normal bundle, but when the library is externalized, this export disappears completely.

Expected:

Running this:

```bash
esbuild index.js --bundle --external:@fluentui/date-time-utilities --format=esm --outdir=dist2
```

Should emit this in the output:

```js
export { DateTime } from '@fluentui/date-time-utilities';
```

**Comment by dzearing** [comment]
Strangely, this throws an error:

`index.js`: `export { DateTime } from './DateTime.js';`
`DateTime.js`: `export * from './path1.js'; export * from './path2.js';`
`path1.js`: `export { DateTime } from '@fluentui/date-time-utilities';`
`path2.js`: `export { DateTime } from '@fluentui/date-time-utilities';`

If instead of the barrel file exporting *, we export * a level deeper, esbuild throws this error:

```
X [ERROR] Ambiguous import "DateTime" has multiple matching exports
```

**Comment by hyrious** [comment]
The problem is caused by esbuild not supporting merging duplicated imports.

**Comment by dzearing** [comment]
Well specifically, duplicated external exports. If the library is not externalized, then esbuild handles it correctly.

**Comment by hinell** [comment]
@dzearing I have the following output for two different runs of the `esbuild v0.15.13`  As you can see FOO got deduped.

 
```js
// $ esbuild a.js --bundle --external:@external --format=esm --outdir=dist/
// cat dist/a.js
// x.js
import { FOO, BAR } from "@external";

// y.js
import { FOO as FOO2, BAZ } from "@external";
export {
  BAR,
  BAZ
};

```

```js
// $ esbuild a.js --format=esm --outdir=dist/
// cat dist/a.js
export * from "./x.js";
export * from "./y.js";

```

for the following INPUT:
#### Inputs
```js
// a.js 
export * from './x.js';
export * from './y.js';

```
```js
// x.js
export { FOO, BAR } from '@external';
```
```js
// y.js
export { FOO, BAZ } from '@external';
```

See attached script for quick reproduction

<details><summary>Quick dirty script to reproduce the issue</summary>


```sh
#!/usr/bin/env bash

TEMPDIR="/tmp/tmp.caNZEeqAHT"

[[ -d $TEMPDIR ]] || mkdir "${TEMPDIR}";
pushd ${TEMPDIR}

[[ -d node_modules/esbuild ]] || pnpm i esbuild;

PATHEXTERNAL=node_modules/@external
[[ -d $PATHEXTERNAL ]] || {
  mkdir $PATHEXTERNAL
}
pushd node_modules/@external
tee index.js <<-EOL
export let FOO = function(){}
export let BAR = "Bar"
export class BAZ {}
EOL
  popd



tee a.js <<EOL
export * from './x.js';
export * from './y.js';
EOL

tee x.js <<EOL
export { FOO, BAR } from '@external';
EOL

tee y.js <<EOL
export { FOO, BAZ } from '@external';
EOL

echo
echo "esbuild version = $(node_modules/.bin/esbuild --version)"

node_modules/.bin/esbuild a.js --bundle --external:@external --format=esm --outdir=dist/
echo "// esbuild a.js --bundle --external:@external --format=esm --outdir=dist/"
echo "// dist/a.js"
cat dist/a.js

node_modules/.bin/esbuild a.js --format=esm --outdir=dist/
echo "// esbuild a.js --format=esm --outdir=dist/"
echo "// dist/a.js"
cat dist/a.js

popd
```

</details>

**Comment by hinell** [comment]
@hyrious It's not missing, it just got renamed by `esbuild` to `FOO2` in the second import. The first source code field is actually entire file.

**Comment by hinell** [comment]
@hyrious It's not missing, it just got renamed by esbuild to FOO2 in the second import. The first source code example is entire file.

**Comment by hyrious** [comment]
@hinell It was renamed but is missing from the exported names (i.e. missing `export { FOO }`). Considering some other guy is importing `FOO` from your `a.js`:

```js
import { FOO } from './dist/a.js'
```

The code will fail because the runtime cannot find the name.

**Comment by hinell** [comment]
@hyrious Yeah, `FOO` is not exported by the `dist/a.js`. It got renamed but not exported eventually. This is a definitely a bug.

**Comment by dzearing** [comment]
Yup @hyrious you got it. I expect `FOO`, `BAR`, and `BAZ` to be exported in that case.

This is actually what we've been observing. Usually, you don't have this case where 2 things emit the same name, so... most of the time you don't notice it. In the above case, the non-duplicated keys (`BAR` and `BAZ`) are emitted, but the duplicate `FOO` is imported twice, and then unused and un-exported.

**Comment by hinell** [comment]
Any progress on this? esbuild should be able to recognize the same module and mrge all its exports requested by the importer.

**Comment by lukeed** [comment]
~> esbuild [REPL](https://esbuild.github.io/try/#YgAwLjE5LjIAewogIGJ1bmRsZTogdHJ1ZSwKICBmb3JtYXQ6ICdlc20nLAogIGV4dGVybmFsOiBbJ0BleHRlcm5hbCddLAp9AGUAZW50cnkuanMAZXhwb3J0ICogZnJvbSAnLi94LmpzJzsKZXhwb3J0ICogZnJvbSAnLi95LmpzJzsAAG5vZGVfbW9kdWxlcy9AZXh0ZXJuYWwvaW5kZXguanMAZXhwb3J0IGxldCBGT08gPSBmdW5jdGlvbigpe30KZXhwb3J0IGxldCBCQVIgPSAiQmFyIgpleHBvcnQgY2xhc3MgQkFaIHt9AAB4LmpzAGV4cG9ydCB7IEZPTywgQkFSIH0gZnJvbSAnQGV4dGVybmFsJzsAAHkuanMAZXhwb3J0IHsgRk9PLCBCQVogfSBmcm9tICdAZXh0ZXJuYWwnOw)
